README.md
# Switchy McPortFace (SMPF)
For want of a better name..

The SMPF client is an application that simply captures switch advertisements broadcast over a local wired network in order to determine information about the switch to which that host is connected, notably the *switch port* that the host is using. Whilst this is possible to do via SNMP or other switch management tools, SMPF aims to accomplish this solely from the hosts themselves. 

Once collected, this information can be output and viewed locally on a host, collected into a local log file, collected into a CSV or sent as POST data to a server. Information that is currently collected so far:
- Switch IP
- Switch name
- Switch MAC address
- Switch port
- Switch port VLAN
- Host IP 
- Host name
- Host MAC address
- Protocol that the data was captured using 

## Libraries
SMPF uses pcap at its core; namely WinPcap*(libpcap 1.0) on Windows or libpcap on Linux systems. This allows for layer 2 frames to be received by the operating system in promiscuous mode. 

*Note: Npcap 'should' be the way forward, but their license terms prohibit actual 'free' use currently, since it is capped at 5 systems.

## Protocols
The SMPF client creates a filter which allows through either CDP and/or LLDP frames. It would be interesting to be able to incorporate Juniper discovery protocol data, or other discovery protocols; however, LLDP is vendor-agnostic and is implemented nearly everywhere. These protocols are not usually used by host operating systems, since they're primarily for switches and routers to build a picture about their neighbours, although they are also leveraged by Fluke testing tools.

Its important to note that not all switches will have LLDP and CDP advertisements enabled, so you're likely out of luck if you don't have access to your network infrastructure in order to enable them (if they aren't already).

Contributions are welcome and I would be really interested to hear from anyone who has Juniper kit, or who wants to implement this or has issues with their own setup


CHANGELOG.md
# Change log - Current version 1.01

## Change history   

### 2019-07-25
**Version 1.01**

* Fixed
  * Tidy up CDP TLVs - CDP dissectors have been tidied
  * Detect if wireless adapters are present and do not listen for packets on these adapters - Checks if the link is wireless or not

* Moved issues and "todo" to the issues list on Github. Future changes will be noted as links to the issues that have been fixed.
* Additional checks to see if the link is virtual or not
* Started code overhaul; created Dissector and POST string classes in preparation for bigger changes
* Removed lots of comments, moved to a legacy branch
* Functionality exists now to log to a file

### 2017-07-20
**Version 1.0**

Nothing significant has changed, this is more a milestone due to its use in the wild. It works and it has a use!
  
### 2017-07-03
Put everything on Github


### 2016-06-22
**Version 0.3**

The client now can be set to output to a log file, rather than to upload data to a server

* To do 
  * Tidy up screen output (\n in the right places) and ensure that debug and release have different behaviour
  * Tidy up LLDP TLVs
  * Tidy up CDP TLVs

### 2016-06-06
**Version 0.2**

cURL is no more! I have sped ahead and managed to cut this out entirely by using WinSock instead of cURL. This means that no extra DLL is needed on Windows, although for consistency on different platforms and SSL support, I may go back to using cURL later and focus on not using separate DLLs.

The server address in the config file now has the page, and the address, as separate items. This just makes it easier when sending the POST request to the server; cURL seemed to split this out (give it the full address, including http://, and it does internal magic to split that up as necessary) but this has to be done manually when working with sockets directly.

* Issues
  * Detect if wireless adapters are present and do not listen for packets on these adapters

* Fixed
  * Find out why CURL can be so slow
  * Replace CURL with native platform socket code


### 2016-05-26
**Version 0.1** 
System is now in use somewhere and tested on ~1500 Windows 7 computers. Version can now jump to 0.1

* Todo
  * Replace CURL with native platform socket code
  
### 2015-10-23
**Version 0.0.4**

An old (but new) update.

CDP packets are now internally formatted to be read as TLVs, based on Cisco's CDP standard (which is hard to find anywhere). Won't bother with any other protocols - if anywhere uses them, though, I can add them in one day.

Added cases for checking capture protocol used (CDP or LLDP) when uploading data and fixed some output formatting.

* Fixed
  * Create TLVs for CDP and other protocols
  * Use TLVs instead of currently "guessing" where data is
  
* Issues
  * It looks like adapters will retain their IP, even if they are inactive and have been disconnected. This can result in disconnected adapters being selected as the interface to listen on, even if they aren't connected.

### 2015-08-24
**Version 0.0.3**

LLDP packets are now internally formatted to be read as TLVs, based on IEEE 802.1AB/2009.

The output for text is now better formatted, too.

* To do
  * Use TLVs instead of currently "guessing" where data is
  * Create TLVs for CDP and other protocols



### 2015-05-02
**Version 0.0.2 (Also really really early)**

Mostly just adding a configuration file. Still lots that needs doing.

A configuration file has now been added. You can specify server address, timeout, whether to use LLDP and/or CDP and what the first octet of the network address is to look for (eg 10, 192, etc). 

All adapters are now checked for the correct adapter. A bug was discovered where, previously, only the first two were ever checked and the program would never exit

* Fixed
  * Create config file reader
  
* To do 
  * Static link libraries


### 2014-12-14
**Version 0.0.1 (Really really early build)**


Currently works by running from command line.


* Issues
  * Find out why CURL is so slow
  * Only checks if a packet is CDP or LLDP, with no distinction for filtering packets destined for phones. Causes incorrect submissions of entries to database.
  
* To do 
  * More debug info
  * Name project
  * Use TLVs instead of currently "guessing" where data is
  * Create config file reader
  * Release build to have no interface at all (automated) unless run from command line
  * Create installer to bundle everything up
  * Create Mac and Linux builds
  * Create makefile and .sln files for distribution
  * Put it all on git or sourceforge
  * FOG module


LanMap Filter/cdp.cpp


//http://www.cisco.com/c/en/us/td/docs/ios-xml/ios/cdp/configuration/15-mt/cdp-15-mt-book/nm-cdp-discover.html


void getDataCDP()
{
	unsigned int count = 0;

	/*
	Destination MAC address - which is octet 0 - 5
	*/
	//%02x, rather than %x, will enforce a mininum field size - so a value of 1 becomes 01 and e becomes 0e. The 0 specifies that it should be 0 padded, the 2 represents the number of 0s.
	printf("\tDestination MAC address:\t %02x:%02x:%02x:%02x:%02x:%02x", packetData[0], packetData[1], packetData[2], packetData[3], packetData[4], packetData[5]);


	/*
	Source MAC address - which is octet 6 - 11
	*/
	printf("\n\tSource MAC address:\t\t %02x:%02x:%02x:%02x:%02x:%02x", packetData[6], packetData[7], packetData[8], packetData[9], packetData[10], packetData[11]);

	/*
	CDP uses Ethernet 1 headers, so bytes 12 and 13 are length
	*/
	printf("\n\tLength:\t\t %i", (packetData[12] << 8 | packetData[13]));

	/*
	CDP uses 802.2 LLC with SNAP headers, sp the next 8 bytes are LLC information, with SNAP extensions for Cisco:
	DSAP (1)
	SSAP (1)
	Control Field (1)

	(snap extension)
	OUI (3)
	Protocol ID (2)

	Bytes 14, 15, 16 are LLC
	Bytes 17, 18, 19 are the OUI (if its 0, then the PID is assumed to be the Ethertype, anything else indicates organisational specific things). Cisco is 0x00000c
	Bytes 20 and 21 are the Protocol ID, which will usually be EtherType. CDP is 2000 (0x2000)


	//https://www.informit.com/library/content.aspx?b=CCNP_Studies_Troubleshooting&seqNum=53
	//https://en.wikipedia.org/wiki/Subnetwork_Access_Protocol

	** CDP **


	For CDP, Bytes 22 and 23 are the version of CDP and the TTL
	24 and 25 are the checksum
	Anything beyond 23 is going to be TLV data. Last 4 octets will be the CRC Checksum, though
	*/

	printf("\n\t Cisco Discovery Protcol version %i\n\t TTL:%i", packetData[22], packetData[23]);


	//http://www.cisco.com/univercd/cc/td/doc/product/lan/trsrb/frames.htm is no longer available apparently, so I cant access it to find out the spec


	count = 26;

	for (count; count < packetHeader->len; )
	{


		unsigned int type = (packetData[count] << 8 | packetData[count + 1]);
		unsigned int length = (packetData[count + 2] << 8 | packetData[count + 3]); //length of the TLV - that includes the type and length itself!
																					//length = length & mask;
		count += 4;
		length -= 4;

		if (length > 500)
		{
			int k = 5;
		}

		unsigned char value[512] = "";

		switch (type)
		{
		case 0x01:
		{
			printf("\n\n\tDevice ID: ");

			memcpy(&systemswName, packetData + count, length);
			slsystemswName = strlen(systemswName);

			for (UINT16 x = 0; x < length; x++)
			{
				value[x] = packetData[count];
				count++;
			}
			printf("%s", value);//, count - length);
			break;
		}
		case 0x02:
		{
			printf("\n\n\tAddress: ");

			//Number of addresses
			long numberOfAddresses = (packetData[count] << 24 | packetData[count] << 16 | packetData[count + 2] << 8 | packetData[count + 3]);
			count += 4;
			length -= 4;
			for (UINT16 x = 0; x < numberOfAddresses; x++)
			{
				//Protocol type
				//Protocol length
				count += 2;

				//This is an IP address
				if (packetData[count] == 0xcc)
				{
					printf("IP Address: ");
					count++;

					//This is how long it is
					int addressLength = (packetData[count] << 8 | packetData[count + 1]);
					count += 2;


					sprintf_s(systemswIP, "%i.%i.%i.%i", packetData[count], packetData[count + 1], packetData[count + 2], packetData[count + 3]);
					printf(systemswIP);
					slsystemswIP = strlen(systemswIP);
					count += addressLength;
				}

			}

			break;
		}
		case 0x03:
		{
			printf("\n\n\tPort-ID: ");

			memcpy(&systemswitchport, packetData + count, length);
			slsystemswitchport = strlen(systemswitchport);

			for (UINT16 x = 0; x < length; x++)
			{
				value[x] = packetData[count];
				count++;
			}
			break;
		}
		case 0x04:
		{
			printf("\n\n\tCapability: ");

			for (UINT16 x = 0; x < length; x++)
			{
				value[x] = packetData[count];
				count++;
			}
			printf("%s", value);//, count - length);
			break;
		}
		case 0x05:
		{
			printf("\n\n\tVersion String: ");

			for (UINT16 x = 0; x < length; x++)
			{
				value[x] = packetData[count];
				count++;
			}
			printf("%s", value);//, count - length);
			break;
		}
		case 0x06:
		{
			printf("\n\n\tPlatform: ");

			memcpy(&systemswMAC, packetData + count, length);
			slsystemswMAC = strlen(systemswMAC);
			for (UINT16 x = 0; x < length; x++)
			{
				value[x] = packetData[count];
				count++;
			}
			printf("%s", value);//, count - length);
			break;
		}
		case 0x07:
		{
			printf("\n\n\tPrefixes: ");

			for (UINT16 x = 0; x < length; x++)
			{
				value[x] = packetData[count];
				count++;
			}
			printf("%s", value);//, count - length);
			break;
		}
		case 0x08:
		{
			printf("\n\n\tProtocol-Hello option: ");

			for (UINT16 x = 0; x < length; x++)
			{
				value[x] = packetData[count];
				count++;
			}
			printf("%s", value);//, count - length);
			break;
		}
		case 0x09:
		{
			printf("\n\n\tVTP Management Domain: ");

			for (UINT16 x = 0; x < length; x++)
			{
				value[x] = packetData[count];
				count++;
			}
			printf("%s", value);//, count - length);
			break;
		}
		case 0x0a:
		{

			printf("\n\n\tNative VLAN ID: ");
			int port_vlan = (packetData[count] << 8 | packetData[count + 1]);

			//printf("VLAN ID: %i", port_vlan);
			_itoa_s(port_vlan, systemvlan, 10);
			slsystemvlan = strlen(systemvlan);

			count += 2;

			printf("%i", port_vlan);
			break;
		}
		case 0x0b:
		{
			printf("\n\n\tDuplex: ");

			for (UINT16 x = 0; x < length; x++)
			{
				value[x] = packetData[count];
				count++;
			}
			printf("%s", value);//, count - length);
			break;
		}
		case 0x0c:
		{
			printf("\n\n\tDONT KNOW LOL");

			for (UINT16 x = 0; x < length; x++)
			{
				value[x] = packetData[count];
				count++;
			}
			printf("%s", value);//, count - length);
			break;
		}
		case 0x0d:
		{
			printf("\n\n\tDONT KNOW LOL");

			for (UINT16 x = 0; x < length; x++)
			{
				value[x] = packetData[count];
				count++;
			}
			printf("%s", value);//, count - length);
			break;
		}
		case 0x0e:
		{
			printf("\n\n\tATA-186 VoIP VLAN request: ");

			for (UINT16 x = 0; x < length; x++)
			{
				value[x] = packetData[count];
				count++;
			}
			printf("%s", value);//, count - length);
			break;
		}
		case 0x0f:
		{
			printf("\n\n\tATA-186 VoIP VLAN assignment: ");

			for (UINT16 x = 0; x < length; x++)
			{
				value[x] = packetData[count];
				count++;
			}
			printf("%s", value);//, count - length);
			break;
		}
		case 0x10:
		{
			printf("\n\n\tPower Consumption: ");

			for (UINT16 x = 0; x < length; x++)
			{
				value[x] = packetData[count];
				count++;
			}
			printf("%s", value);//, count - length);
			break;
		}
		case 0x11:
		{
			printf("\n\n\tMTU: ");

			for (UINT16 x = 0; x < length; x++)
			{
				value[x] = packetData[count];
				count++;
			}
			printf("%s", value);//, count - length);
			break;
		}
		case 0x12:
		{
			printf("\n\n\tAVVID Trust Bitmap: ");

			for (UINT16 x = 0; x < length; x++)
			{
				value[x] = packetData[count];
				count++;
			}
			printf("%s", value);//, count - length);
			break;
		}
		case 0x13:
		{
			printf("\n\n\tAVVID Untrusted Ports CoS: ");

			for (UINT16 x = 0; x < length; x++)
			{
				value[x] = packetData[count];
				count++;
			}
			printf("%s", value);//, count - length);
			break;
		}
		case 0x14:
		{
			printf("\n\n\tSystem Name: ");

			for (UINT16 x = 0; x < length; x++)
			{
				value[x] = packetData[count];
				count++;
			}
			printf("%s", value);//, count - length);
			break;
		}
		case 0x15:
		{
			printf("\n\n\tSystem Object ID (Not Decoded): ");

			for (UINT16 x = 0; x < length; x++)
			{
				value[x] = packetData[count];
				count++;
			}
			printf("%s", value);//, count - length);
			break;
		}
		case 0x16:
		{
			printf("\n\n\tManagement Addresses: ");

			for (UINT16 x = 0; x < length; x++)
			{
				value[x] = packetData[count];
				count++;
			}
			printf("%s", value);//, count - length);
			break;
		}
		case 0x17:
		{
			printf("\n\n\tPhysical Location: ");

			for (UINT16 x = 0; x < length; x++)
			{
				value[x] = packetData[count];
				count++;
			}
			printf("%s", value);//, count - length);
			break;
		}
		case 0x18:
		{
			printf("\n\n\tUNKNOWN: ");

			for (UINT16 x = 0; x < length; x++)
			{
				value[x] = packetData[count];
				count++;
			}
			printf("%s", value);//, count - length);
			break;
		}
		case 0x19:
		{
			printf("\n\n\tUNKNOWN: ");

			for (UINT16 x = 0; x < length; x++)
			{
				value[x] = packetData[count];
				count++;
			}
			printf("%s", value);//, count - length);
			break;
		}
		case 0x1a:
		{
			printf("\n\n\tPower Available: ");

			for (UINT16 x = 0; x < length; x++)
			{
				value[x] = packetData[count];
				count++;
			}
			printf("%s", value);//, count - length);
			break;
		}
		case 0x1b:
		{
			printf("\n\n\tUNKNOWN: ");

			for (UINT16 x = 0; x < length; x++)
			{
				value[x] = packetData[count];
				count++;
			}
			printf("%s", value);//, count - length);
			break;
		}
		case 0x1c:
		{
			printf("\n\n\tUNKNOWN: ");

			for (UINT16 x = 0; x < length; x++)
			{
				value[x] = packetData[count];
				count++;
			}
			printf("%s", value);//, count - length);
			break;
		}
		case 0x1d:
		{
			printf("\n\n\tEnergyWise: ");

			for (UINT16 x = 0; x < length; x++)
			{
				value[x] = packetData[count];
				count++;
			}
			printf("%s", value);//, count - length);
			break;
		}
		case 0x1e:
		{
			printf("\n\n\tUNKNOWN: ");

			for (UINT16 x = 0; x < length; x++)
			{
				value[x] = packetData[count];
				count++;
			}
			printf("%s", value);//, count - length);
			break;
		}
		case 0x1f:
		{
			printf("\n\n\tSpare PoE: ");

			for (UINT16 x = 0; x < length; x++)
			{
				value[x] = packetData[count];
				count++;
			}
			printf("%s", value);//, count - length);
			break;
		}
		default:
		{
			printf("\n\n\t *** Unknown *** value = %s", value);
			break;
		}
		}
	}
}